buildpacktoml_test.go

```diff
@@ -0,0 +1,729 @@
+					it("should error", func() {
+						env.EXPECT().WithPlatform(platformDir).Return(append(os.Environ(), "TEST_ENV=Av1"), nil)
+						mkfile(t,
+							"[[unmet]]\n"+
+								`name = "unknown-dep"`+"\n",
+							filepath.Join(appDir, "build-A-v1.toml"),
+						)
+						_, err := bpTOML.Build(lifecycle.BuildpackPlan{}, config)
+						h.AssertNotNil(t, err)
+						expected := "must match a requested dependency"
+						h.AssertStringContains(t, err.Error(), expected)
+					})
+				})
+			})
+		})
+
+		when("buildpack api = 0.2", func() {
+			it.Before(func() {
+				bpTOML.API = "0.2"
+				env.EXPECT().WithPlatform(platformDir).Return(append(os.Environ(), "TEST_ENV=Av1"), nil)
+			})
+
+			it("should convert metadata version to top level version in the buildpack plan", func() {
+				bpPlan := lifecycle.BuildpackPlan{
+					Entries: []lifecycle.Require{
+						{
+							Name:     "some-dep",
+							Metadata: map[string]interface{}{"version": "v1"},
+						},
+					},
+				}
+
+				_, err := bpTOML.Build(bpPlan, config)
+				if err != nil {
+					t.Fatalf("Unexpected error:\n%s\n", err)
+				}
+
+				testPlan(t,
+					[]lifecycle.Require{
+						{
+							Name:     "some-dep",
+							Version:  "v1",
+							Metadata: map[string]interface{}{"version": "v1"},
+						},
+					},
+					filepath.Join(appDir, "build-plan-in-A-v1.toml"),
+				)
+			})
+		})
+
+		when("buildpack api < 0.5", func() {
+			it.Before(func() {
+				bpTOML.API = "0.4"
+			})
+
+			when("building succeeds", func() {
+				it.Before(func() {
+					env.EXPECT().WithPlatform(platformDir).Return(append(os.Environ(), "TEST_ENV=Av1"), nil)
+				})
+
+				it("should get bom entries and unmet requires from the buildpack plan", func() {
+					bpPlan := lifecycle.BuildpackPlan{
+						Entries: []lifecycle.Require{
+							{
+								Name:    "some-deprecated-bp-dep",
+								Version: "v1", // top-level version is deprecated in buildpack API 0.3
+							},
+							{
+								Name:    "some-deprecated-bp-replace-version-dep",
+								Version: "some-version-orig", // top-level version is deprecated in buildpack API 0.3
+							},
+							{
+								Name:    "some-deprecated-bp-add-version-dep",
+								Version: "v2", // top-level version is deprecated in buildpack API 0.3
+							},
+							{
+								Name:    "some-deprecated-bp-move-version-dep",
+								Version: "v3", // top-level version is deprecated in buildpack API 0.3
+							},
+							{
+								Name:     "some-dep",
+								Metadata: map[string]interface{}{"version": "v1"},
+							},
+							{
+								Name:     "some-replace-version-dep",
+								Metadata: map[string]interface{}{"version": "some-version-orig"},
+							},
+							{
+								Name: "some-unmet-dep",
+							},
+						},
+					}
+
+					mkfile(t,
+						"[[entries]]\n"+
+							`name = "some-deprecated-bp-dep"`+"\n"+
+							`version = "v1"`+"\n"+
+							"[[entries]]\n"+
+							`name = "some-deprecated-bp-replace-version-dep"`+"\n"+
```

Comment:
This is onerous... even as I was writing this code I felt that I was going too far. It may be enough to assert something like "BOM is an exact reflection of output buildpack plan" or something like that.

Reply:
I like the `BOM is an exact reflection of output buildpack plan` idea